### PR TITLE
Save total with and without tax in invoiceinline

### LINKED ISSUE #366: Line total without tax and tax is different on FB and LODO
The invoice which came through XML to FB was downloaded to LODO, and it had different line total without tax and line total tax.

This is the example:
https://fakturabank.no/invoices/204616
Ocean Cafe Bar AS U6934

Data sent to LODO should be same as the one which we received with XML in FB.

### PR DESCRIPTION
**Pull request by Baxxx**
Closes #366

### CODE REVIEW

File: db/changes/160_add_total_without_tax_to_invoicein.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE invoiceinline ADD COLUMN TotalWithoutTax decimal(16,2), ADD COLUMN TotalWithTax decimal(16,2);
+
+UPDATE invoiceinline SET TotalWithoutTax = (IF(ForeignCurrencyID is null, UnitCustPrice * QuantityDelivered, ForeignAmount - TaxAmount) + IFNULL((SELECT SUM(IF(ChargeIndicator, Amount, -Amount)) FROM invoicelineallowancecharge WHERE AllowanceChargeType = 'line' AND InvoiceType = 'in' AND InvoiceLineID = invoiceinline.LineID), 0));
```
Comment (Baxxx) [inline db/changes/160_add_total_without_tax_to_invoicein.sql]: Always use `UnitCustPrice * QuantityDelivered`

File: db/changes/160_add_total_without_tax_to_invoicein.sql [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+UPDATE invoiceinline SET TaxAmount = TotalWithoutTax * Vat / 100;
```
Comment (sls-lnluksa) [inline db/changes/160_add_total_without_tax_to_invoicein.sql]: Calculate TotalWithTax first, with `TotalWithoutTax * (1+Vat/100)`

File: modules/invoicein/model/invoicein.class.php [hunk 3]
```diff
@@ -204,6 +204,84 @@ function addVATPercentToAllowanceCharge(&$args) {
...
+                        } else {
+                            $line_allowance_charge_sum -= $amount;
+                        }
+                    }
+                }
+            }
+
+            if($fields["TotalWithTax"]) { // this is in current invoice's currency
+                if(!$fields["TotalWithoutTax"]) {
+                    $fields["TotalWithoutTax"] = $fields["TotalWithTax"] / (1 + $fields["Vat"] / 100);
```
Comment (sls-lnluksa) [inline modules/invoicein/model/invoicein.class.php]: Round all amounts to two decimal

File: modules/invoicein/model/invoicein.class.php [hunk 4]
```diff
@@ -204,6 +204,84 @@ function addVATPercentToAllowanceCharge(&$args) {
...
+                if($fields["TotalWithoutTax"]) {
+                    if (!$fields["QuantityDelivered"] && $fields["UnitCustPrice"]) {
+                        $fields["QuantityDelivered"] = ($fields["TotalWithoutTax"] - $line_allowance_charge_sum) / $fields["UnitCustPrice"];
+                    } else if($fields["QuantityDelivered"]) {
+                        $fields["UnitCustPrice"] = ($fields["TotalWithoutTax"] - $line_allowance_charge_sum) / $fields["QuantityDelivered"];
+                    }
+                } else {
+                    $fields["TotalWithoutTax"] = $fields["QuantityDelivered"] * $fields["UnitCustPrice"] + $line_allowance_charge_sum;
+                }
+                $fields["TaxAmount"] = $fields["TotalWithoutTax"] * $fields["Vat"] / 100;
```
Comment (sls-lnluksa) [inline modules/invoicein/model/invoicein.class.php]: Calculate first `TotalWithTax` than `TaxAmount = TotalWithTax - TotalWithoutTax`

File: modules/invoicein/model/invoicein.class.php [hunk 5]
```diff
@@ -204,6 +204,84 @@ function addVATPercentToAllowanceCharge(&$args) {
...
+                $fields["TotalWithTax"] = $fields["TotalWithoutTax"] + $fields["TaxAmount"];
+            }
+
+            $fields["TaxAmount"] = $fields["TotalWithoutTax"] * $fields["Vat"] / 100;
```
Comment (sls-lnluksa) [inline modules/invoicein/model/invoicein.class.php]: This is excess. Remove.

**Comment by sls-lnluksa** [comment]
Check PDF invoice. Use new values in PDF and pay attention to allowance/charges on line level.

**Review by sls-lnluksa — APPROVED** [review]
Final review with: @stoivo 

⚠️  BEFORE DEPLOY
@stoivo @agvald @hildegunn95 
Test downloading, and edit downloaded invoice.
Use example from issue and more...

File: modules/invoicein/model/invoicein.class.php [hunk 6]
```diff
@@ -204,6 +204,83 @@ function addVATPercentToAllowanceCharge(&$args) {
...
+        foreach ($args as $arg_key => $arg_val) {
+            preg_match("/^invoiceinline_.*_(\d*)$/", $arg_key, $matches);
+            $invoiceinline_id = $matches[1];
+            if($invoiceinline_id && !in_array($invoiceinline_id, $invoiceinline_ids)) {
+                $invoiceinline_ids[] = $invoiceinline_id;
+            }
+        }
+
+        $total_invoice_sum = 0;
+        $line_allowance_charge_sum = 0;
```
Comment (Baxxx) [inline modules/invoicein/model/invoicein.class.php]: This should be inside of foreach

File: modules/invoicein/model/invoicein.class.php [hunk 7]
```diff
@@ -204,6 +204,83 @@ function addVATPercentToAllowanceCharge(&$args) {
...
+            foreach ($args as $arg_key => $arg_val) {
+                if(preg_match("/^invoiceinline_(.*)_".$line_id."$/", $arg_key, $matches)) {
+                    $args[$arg_key] = $_lib['convert']->Amount($arg_val);
+                    if(in_array($matches[1], array("TotalWithoutTax", "UnitCustPrice", "QuantityDelivered", "TotalWithTax", "Vat"))) {
+                        $fields[$matches[1]] = (float)$args["invoiceinline_$matches[1]_$line_id"];
+                    }
+                }
+                if(preg_match("/^invoicelineallowancecharge_(.*)_(.*)$/", $arg_key, $matches)) {
+                    $invoicelineallowancecharge_id = $matches[2];
+                    if($matches[1] == "InvoiceLineID" && $args["invoicelineallowancecharge_AllowanceChargeType_$invoicelineallowancecharge_id"] == "line") {
```
Comment (Baxxx) [inline modules/invoicein/model/invoicein.class.php]: Also, this allowancecharge should belong to this line

Comment (Baxxx) [reply modules/invoicein/model/invoicein.class.php]: if($args["invoicelineallowancecharge_InvoiceLineID_$invoicelineallowancecharge_id"] == $line_id)

File: modules/invoicein/model/invoicein.class.php [hunk 8]
```diff
@@ -483,39 +561,13 @@ public function journal() {
...
-                              $TotalPrice = round(($TotalPrice + $TaxAmount), 2);
-                            }
-                            $invoice_line_sum += $TotalPrice;
-                        } else {
-                            $invoice_line_sum += $TotalPrice;
-                        }
+                        $TotalPrice         = round($line->TotalWithTax, 2);
+                        $TotalForeignPrice  = round($line->ForeignAmount, 2);
+
+                        $invoice_line_sum += round($line->TotalWithTax, 2);
```
Comment (Baxxx) [inline modules/invoicein/model/invoicein.class.php]: Use $TotalPrice

File: modules/invoicein/model/invoicein.class.php [hunk 9]
```diff
@@ -527,7 +579,9 @@ public function journal() {
                             $VoucherH['voucher_AmountIn']   = 0;
                         }
 
-                        if ($VoucherH['voucher_ForeignCurrencyID'] != '') $VoucherH['voucher_ForeignAmount']   = abs($TotalForeignPrice);
+                        $VoucherH['voucher_ForeignAmount']   = abs($TotalForeignPrice);
+
+                        $in_or_out = ($line->TotalWithTax > 0 ? 'in' : 'out');
```
Comment (Baxxx) [inline modules/invoicein/model/invoicein.class.php]: Move this down, above the `insert_voucher_line`

**Comment by hildegunn95** [comment]
Tested on invoice with fakturabank id 203422
skjermbilde 2017-02-01 kl 08 35 07

**Comment by sls-lnluksa** [comment]
- [ ] Test line calculations on invoicein.edit page. Amounts that we want to recalculate we should delete from page and then press save.

File: modules/invoicein/model/invoicein.class.php [hunk 10]
```diff
@@ -204,6 +204,85 @@ function addVATPercentToAllowanceCharge(&$args) {
...
+            foreach ($args as $arg_key => $arg_val) {
+                if(preg_match("/^invoiceinline_(.*)_".$line_id."$/", $arg_key, $matches)) {
+                    $args[$arg_key] = $_lib['convert']->Amount($arg_val);
+                    if(in_array($matches[1], array("TotalWithoutTax", "UnitCustPrice", "QuantityDelivered", "TotalWithTax", "Vat"))) {
+                        $fields[$matches[1]] = (float)$args["invoiceinline_$matches[1]_$line_id"];
+                    }
+                }
+                if(preg_match("/^invoicelineallowancecharge_(.*)_(.*)$/", $arg_key, $matches)) {
+                    $invoicelineallowancecharge_id = $matches[2];
+                    if( $matches[1] == "InvoiceLineID" &&
```
Comment (Baxxx) [inline modules/invoicein/model/invoicein.class.php]: Merge this with regex above

**Review by stoivo — APPROVED** [review]

**Pull request merged**